atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -658,6 +638,176 @@ public int compare(Entry<Cell, Long> entry1, Entry<Cell, Long> entry2) {
+                Long overflowId = hasOverflow ? row.getLongObject("overflow") : null;
+                if (overflowId == null) {
+                    Value value = Value.create(row.getBytes("val"), row.getLong("ts"));
+                    Value oldValue = values.put(cell, value);
+                    if (oldValue != null && oldValue.getTimestamp() > value.getTimestamp()) {
+                        values.put(cell, oldValue);
+                    }
+                } else {
+                    OverflowValue ov = new OverflowValue(row.getLong("ts"), overflowId);
+                    OverflowValue oldOv = overflowValues.put(cell, ov);
+                    if (oldOv != null && oldOv.ts > ov.ts) {
+                        overflowValues.put(cell, oldOv);
+                    }
+                }
+            }
+        } finally {
+            iter.close();
+        }
+        fillOverflowValues(table, overflowValues, values);
+
+        Map<byte[], List<Map.Entry<Cell, Value>>> results = Maps.newHashMapWithExpectedSize(rows.size());
+        for (Entry<Sha256Hash, List<Cell>> e : cellsByRow.entrySet()) {
+            List<Map.Entry<Cell, Value>> fullResults = Lists.newArrayListWithExpectedSize(e.getValue().size());
+            for (Cell c : e.getValue()) {
+                fullResults.add(Iterables.getOnlyElement(ImmutableMap.of(c, values.get(c)).entrySet()));
+            }
+            results.put(hashesToBytes.get(e.getKey()), fullResults);
+        }
+        return results;
+    }
+
+    private void fillOverflowValues(DbReadTable table,
+                                    Map<Cell, OverflowValue> overflowValues,
+                                    @Output Map<Cell, Value> values) {
+        for (Iterator<Entry<Cell, OverflowValue>> entryIter = overflowValues.entrySet().iterator(); entryIter.hasNext(); ) {
+            Entry<Cell, OverflowValue> entry = entryIter.next();
+            Value value = values.get(entry.getKey());
+            if (value != null && value.getTimestamp() > entry.getValue().ts) {
+                entryIter.remove();
+            }
+        }
+        if (!overflowValues.isEmpty()) {
+            Map<Long, byte[]> resolvedOverflowValues = Maps.newHashMapWithExpectedSize(overflowValues.size());
+            ClosableIterator<AgnosticLightResultRow> overflowIter = table.getOverflow(overflowValues.values());
+            try {
+                while (overflowIter.hasNext()) {
+                    AgnosticLightResultRow row = overflowIter.next();
+                    // QA-94468 LONG RAW typed columns ("val" in this case) must be retrieved first from the result set
+                    // see https://docs.oracle.com/cd/B19306_01/java.102/b14355/jstreams.htm#i1007581
+                    byte[] val = row.getBytes("val");
+                    long id = row.getLong("id");
+                    resolvedOverflowValues.put(id, val);
+                }
+            } finally {
+                overflowIter.close();
```

Comment:
Could use try with resources to autoclose and get suppressed exceptions on failure